src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/BlobUtilities.cs

```diff
@@ -0,0 +1,282 @@
+        {
+            fixed (byte* ptr = &buffer[start])
+            {
+                *(double*)ptr = value;
+            }
+        }
+
+        public static void WriteSingle(this byte[] buffer, int start, float value)
+        {
+            fixed (byte* ptr = &buffer[start])
+            {
+                *(float*)ptr = value;
+            }
+        }
+
+        public static void WriteByte(this byte[] buffer, int start, byte value)
+        {
+            // Perf: The compiler emits a check when pinning the buffer. It's thus not worth doing so.
+            buffer[start] = value;
+        }
+
+        public static void WriteUInt16(this byte[] buffer, int start, ushort value)
+        {
+            fixed (byte* ptr = &buffer[start])
+            {
+                *(ushort*)ptr = value;
+            }
+        }
+
+        public static void WriteUInt16BE(this byte[] buffer, int start, ushort value)
+        {
+            fixed (byte* ptr = &buffer[start])
+            {
+                unchecked
+                {
+                    ptr[0] = (byte)(value >> 8);
+                    ptr[1] = (byte)value;
+                }
+            }
+        }
+
+        public static void WriteUInt32BE(this byte[] buffer, int start, uint value)
+        {
+            fixed (byte* ptr = &buffer[start])
+            {
+                unchecked
+                {
+                    ptr[0] = (byte)(value >> 24);
+                    ptr[1] = (byte)(value >> 16);
+                    ptr[2] = (byte)(value >> 8);
+                    ptr[3] = (byte)value;
+                }
+            }
+        }
+
+        public static void WriteUInt32(this byte[] buffer, int start, uint value)
+        {
+            fixed (byte* ptr = &buffer[start])
+            {
+                *(uint*)ptr = value;
+            }
+        }
+
+        public static void WriteUInt64(this byte[] buffer, int start, ulong value)
+        {
+            fixed (byte* ptr = &buffer[start])
+            {
+                *(ulong*)ptr = value;
+            }
+        }
+
+        public const int SizeOfSerializedDecimal = sizeof(byte) + 3 * sizeof(uint);
+
+        public static void WriteDecimal(this byte[] buffer, int start, decimal value)
+        {
+            bool isNegative;
+            byte scale;
+            uint low, mid, high;
+            value.GetBits(out isNegative, out scale, out low, out mid, out high);
+
+            fixed (byte* ptr = &buffer[start])
+            {
+                *ptr = (byte)(scale | (isNegative ? 0x80 : 0x00));
+                *(uint*)(ptr + 1) = low;
+                *(uint*)(ptr + 5) = mid;
+                *(uint*)(ptr + 9) = high;
+            }
+        }
+
+        // TODO: Use UTF8Encoding https://github.com/dotnet/corefx/issues/2217
+        public static void WriteUTF8(this byte[] buffer, int start, char* charPtr, int charCount, int byteCount, bool allowUnpairedSurrogates)
```

Comment:
What are the implications of the *reader* not supporting unpaired surrogates? E.g. Existing ReadSerializedString is hard-wired to UTF8Encoding IIRC.